docs/kernel_providers.rst

```diff
@@ -0,0 +1,173 @@
+                'argv': ['oblong-kernel'],
+            }
+
+        def launch(self, name, cwd=None, launch_params=None):
+            if name == 'standard':
+                return my_launch_method(cwd=cwd, launch_params=launch_params,
+                                        kernel_cmd=['oblong-kernel'],
+                                        extra_env={'ROUNDED': '0'})
+            elif name == 'rounded':
+                return my_launch_method(cwd=cwd, launch_params=launch_params,
+                                        kernel_cmd=['oblong-kernel'],
+                                        extra_env={'ROUNDED': '1'})
+            else:
+                raise ValueError("Unknown kernel %s" % name)
+
+You would then register this with an *entry point*. In your ``setup.py``, put
+something like this::
+
+    setup(...
+        entry_points = {
+        'jupyter_kernel_mgmt.kernel_type_providers' : [
+            # The name before the '=' should match the id attribute
+            'oblong = oblong_provider:OblongKernelProvider',
+        ]
+    })
+
+Finding kernel types
+====================
+
+To find and start kernels in client code, use
+:class:`KernelFinder <jupyter_kernel_mgmt.discovery.KernelFinder>`. This uses multiple kernel
+providers to find available kernels. Like a kernel provider, it has methods
+``find_kernels`` and ``launch``. The kernel names it works
+with have the provider ID as a prefix, e.g. ``oblong/rounded`` (from the example
+above).
+
+::
+
+    from jupyter_kernel_mgmt.discovery import KernelFinder
+    kf = KernelFinder.from_entrypoints()
+
+    ## Find available kernel types
+    for name, attributes in kf.find_kernels():
+        print(name, ':', attributes['display_name'])
+    # oblong/standard : Oblong (standard)
+    # oblong/rounded : Oblong(rounded)
+    # ...
+
+    ## Start a kernel by name
+    connect_info, manager = kf.launch('oblong/standard')
+
+    # TODO - Finish example with IOLoopKernelClient, etc.
+
+Included kernel providers
+=========================
+
+``jupyter_kernel_mgmt`` includes two kernel providers in its distribution.
+
+1. :class:`KernelSpecProvider <jupyter_kernel_mgmt.discovery.KernelSpecProvider>` handles the discovery and launch
+of most existing kernelspec-based kernels that exist today.  If a provider id is not provided
+when calling :meth:`KernelFinder.launch(name) <.KernelFinder.launch>`, KernelFinder will assume its associated with
+a kernelspec provider and use :class:`KernelSpecProvider <jupyter_kernel_mgmt.discovery.KernelSpecProvider>` to
+perform the launch.
+
+2. :class:`IPykernelProvider <jupyter_kernel_mgmt.discovery.IPykernelProvider>` handles the discover and launch of any
+IPython kernel that is located in the executing python's interpreter.  For example, if the
+application is running in a virtual Python environment, this provider identifies any IPython
+kernels that are local to that environment and may not be identified by the path algorithm
```

Comment:
To be clear, it only ever identifies 0 or 1 kernel, depending on whether ipykernel is importable, never kernels plural.

It's meant to be a coherent replacement of jupyter_client's special-case fallback if there's no `python3` kernelspec.